core/iwasm/common/wasm_c_api.c

```diff
@@ -1651,72 +1651,44 @@ wasm_trap_new_internal(WASMModuleInstanceCommon *inst_comm_rt,
         return NULL;
     }
 
-#if WASM_ENABLE_INTERP != 0
-    if (inst_comm_rt->module_type == Wasm_Module_Bytecode) {
-        error_info = wasm_get_exception((WASMModuleInstance *)inst_comm_rt);
-    }
-#endif
-
-#if WASM_ENABLE_AOT != 0
-    if (inst_comm_rt->module_type == Wasm_Module_AoT) {
-        error_info = aot_get_exception((AOTModuleInstance *)inst_comm_rt);
-    }
-#endif
-
-    /*
-     * a wrong combination of module filetype and compilation flags
-     * also leads to below branch
-     */
-    if (!error_info && !(error_info = default_error_info)) {
-        return NULL;
-    }
-
     if (!(trap = malloc_internal(sizeof(wasm_trap_t)))) {
         return NULL;
     }
 
-    if (!(trap->message = malloc_internal(sizeof(wasm_byte_vec_t)))) {
-        goto failed;
-    }
+    /* fill in message */
+    if (strlen(error_info) > 0) {
+        if (!(trap->message = malloc_internal(sizeof(wasm_byte_vec_t)))) {
+            goto failed;
+        }
 
-    wasm_name_new_from_string_nt(trap->message, error_info);
-    if (strlen(error_info) && !trap->message->data) {
-        goto failed;
+        wasm_name_new_from_string_nt(trap->message, error_info);
+        if (!trap->message->data) {
+            goto failed;
+        }
     }
 
+    /* fill in frames */
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
-#if WASM_ENABLE_INTERP != 0
-    if (inst_comm_rt->module_type == Wasm_Module_Bytecode) {
-        trap->frames = ((WASMModuleInstance *)inst_comm_rt)->frames;
-    }
-#endif
-
-#if WASM_ENABLE_AOT != 0
-    if (inst_comm_rt->module_type == Wasm_Module_AoT) {
-        trap->frames = ((AOTModuleInstance *)inst_comm_rt)->frames;
-    }
-#endif
-#endif /* WASM_ENABLE_DUMP_CALL_STACK != 0 */
-
-    /* allow a NULL frames list */
-    if (!trap->frames) {
-        return trap;
-    }
+    trap->frames = ((WASMModuleInstance *)inst_comm_rt)->frames;
 
-    if (!(instances = singleton_engine->stores->data[0]->instances)) {
-        goto failed;
-    }
+    if (trap->frames) {
+        /* fill in instances */
+        instances = store->instances;
+        bh_assert(instances != NULL);
 
-    for (i = 0; i < instances->num_elems; i++) {
-        if (instances->data[i]->inst_comm_rt == inst_comm_rt) {
-            frame_instance = instances->data[i];
-            break;
+        for (i = 0; i < instances->num_elems; i++) {
+            if (instances->data[i]->inst_comm_rt == inst_comm_rt) {
+                frame_instance = instances->data[i];
+                break;
+            }
         }
-    }
 
-    for (i = 0; i < trap->frames->num_elems; i++) {
-        (((wasm_frame_t *)trap->frames->data) + i)->instance = frame_instance;
+        for (i = 0; i < trap->frames->num_elems; i++) {
+            (((wasm_frame_t *)trap->frames->data) + i)->instance =
+                frame_instance;
+        }
     }
+#endif /* WASM_ENABLE_DUMP_CALL_STACK != 0 */
```

Comment:
In the original code, this line is in L1699 (old code), and normally it should be move to L1673 (new code), now the macro wraps more codes, is it OK?